csp/utils.py

```diff
@@ -6,85 +6,178 @@
+    policies = getattr(settings, 'CSP_POLICIES', defaults.POLICIES)
+    definitions = csp_definitions_update({}, defaults.POLICY_DEFINITIONS)
+    custom_definitions = getattr(
+        settings,
+        'CSP_POLICY_DEFINITIONS',
+        {'default': {}},
+    )
+     # Technically we're modifying Django settings here,
+     # but it shouldn't matter since the end result of either will be the same
+    deprecation._handle_legacy_settings(custom_definitions)
+    for name, csp in custom_definitions.items():
+        definitions[name].update(csp)
+    # TODO: Error handling
+    # TODO: Remove in October 2020 when ordered dicts are the default
+    return OrderedDict(
+        (name, definitions[name]) for name in policies
+    )
+
+
+def _normalize_config(config, key='default'):
+    """
+    Permits the config to be a single policy, which will be returned under the
+    'default' key by default.
+    """
+    if config is None:
+        return {}
+    if not config:
+        return config
+
+    if not isinstance(next(iter(config.values())), dict):
+        return {'default': config}
+    return config
+
+
+def build_policy(
+    config=None,
+    update=None,
+    replace=None,
+    nonce=None,
+    order=None,
+):
+    """
+    Builds the policy from the settings as a list of tuples:
+    (policy_string<str>, report_only<bool>)
+    """
     if config is None:
         config = from_settings()
+    else:
+        config = _normalize_config(config)
         # Be careful, don't mutate config as it could be from settings
+    update = _normalize_config(update)
+    replace = _normalize_config(replace)
+
+    policies = OrderedDict()
+    for name, policy in config.items():
+        new_policy = _replace_policy(policy, replace.get(name, {}))
+        update_policy = update.get(name)
+        if update_policy is not None:
+            _update_policy(new_policy, update_policy)
+        policies[name] = new_policy
+    _append_policies(policies, update, config)
+
+    if order:  # empty order not permitted: use csp_exempt instead
+        policies = (policies[index_or_name] for index_or_name in order)
+    else:
+        policies = policies.values()
 
-    update = update if update is not None else {}
-    replace = replace if replace is not None else {}
-    csp = {}
+    return [_compile_policy(csp, nonce=nonce) for csp in policies]
 
-    for k in set(chain(config, replace)):
-        if k in replace:
-            v = replace[k]
+
+def _append_policies(policies, append, config=None):
+    if config is None:
+        config = {}
+    for key, append_csp in append.items():
+        if key in policies:
+            continue
+
+        if key in config and not append_csp:
+            csp = config[key].copy()
         else:
-            v = config[k]
-        if v is not None:
-            v = copy.copy(v)
-            if not isinstance(v, (list, tuple)):
-                v = (v,)
-            csp[k] = v
+            # TODO: design decision
```

Comment:
Design Decision: Do we need this append functionality or should we just error out if the policy we're updating doesn't exist?  I guess the issue is handling legacy calls?

Reply:
I've simplified this significantly.  csp_update and csp_replace no longer support appending.  I added a new csp_append decorator, but not sure if it's desirable.

Design Decision:  csp_update and csp_append using the old capitalized kwarg style only apply to the "default" policy by default, do we want them to apply to *all* policies?  I think not because it could be a surprise that bites someone.